src/lib.rs

```diff
@@ -694,7 +856,26 @@ impl<T> ThinVec<T> {
             (*ptr).set_cap(new_cap);
             self.ptr = NonNull::new_unchecked(ptr);
         } else {
-            self.ptr = header_with_capacity::<T>(new_cap);
+            let mut new_header = header_with_capacity::<T>(new_cap);
+
+            // If we get here and have a non-zero len, then we must be handling
+            // a gecko auto array, and we have items in a stack buffer. We shouldn't
+            // free it, but we should memcopy the contents out of it and mark it as empty.
+            //
+            // T is assumed to have a trivial move constructor, as this is required
+            // for Rust compatibility anyway. Furthermore, we assume C++ won't try
```

Comment:
Leaning on https://github.com/Gankra/thin-vec/pull/18#discussion_r466893153, I still don't think this is correct. `ThinVec` is instantiated as `ThinVec<RefPtr<nsIKeyValuePair>>` and `ThinVec<nsCString>`, e.g., and neither `RefPtr<nsIKeyValuePair>` nor `nsCString` have a trivial move constructor.

Can you elaborate more what "required for Rust compatibility" means?